src/app/clusters/basic/basic.cpp

```diff
@@ -62,41 +62,251 @@ CHIP_ERROR BasicAttrAccess::Read(const ConcreteReadAttributePath & aPath, Attrib
+            productLabel[0] = '\0';
+            status          = CHIP_NO_ERROR;
+        }
+
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(chip::CharSpan(productLabel, strnlen(productLabel, kMaxLen)));
+        }
+        break;
+    }
+
+    case SerialNumber::Id: {
+        constexpr size_t kMaxLen             = DeviceLayer::ConfigurationManager::kMaxSerialNumberLength;
+        char serialNumberString[kMaxLen + 1] = { 0 };
+        status                               = ConfigurationMgr().GetSerialNumber(serialNumberString, sizeof(serialNumberString));
+
+        // TODO: Remove defaulting once proper runtime defaulting of unimplemented factory data is done
+        if (status == CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND || status == CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE)
+        {
+            serialNumberString[0] = '\0';
+            status                = CHIP_NO_ERROR;
+        }
+
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(chip::CharSpan(serialNumberString, strnlen(serialNumberString, kMaxLen)));
+        }
+        break;
+    }
+
+    case UniqueID::Id: {
+        constexpr size_t kMaxLen   = DeviceLayer::ConfigurationManager::kMaxUniqueIDLength;
+        char uniqueId[kMaxLen + 1] = { 0 };
+        status                     = ConfigurationMgr().GetUniqueId(uniqueId, sizeof(uniqueId));
+
+        // TODO: Remove defaulting once proper runtime defaulting of unimplemented factory data is done
+        if (status == CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND || status == CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE)
         {
-            err = aEncoder.Encode(chip::CharSpan("XX", strlen("XX")));
+            uniqueId[0] = '\0';
+            status      = CHIP_NO_ERROR;
         }
-        else
+
+        if (status == CHIP_NO_ERROR)
         {
-            err = aEncoder.Encode(chip::CharSpan(location, strlen(location)));
+            status = aEncoder.Encode(chip::CharSpan(uniqueId, strnlen(uniqueId, kMaxLen)));
         }
+        break;
+    }
+
+    default:
+        // We did not find a processing path, the caller will delegate elsewhere.
+        status = CHIP_ERROR_NOT_FOUND;
+        break;
     }
-    else
+
+    if (status == CHIP_ERROR_NOT_FOUND)
     {
-        err = aEncoder.Encode(chip::CharSpan("XX", strlen("XX")));
+        // Somehow attribute support not handled by custom logic. Until all such
+        // paths are resolved, allow fallback to attribute store or other heuristics
+        // which may be done by callers of this method.
+        status = CHIP_NO_ERROR;
     }
 
-    return err;
+    return status;
+}
+
+CHIP_ERROR BasicAttrAccess::ReadLocation(AttributeValueEncoder & aEncoder)
+{
+    constexpr size_t kMaxLen   = DeviceLayer::ConfigurationManager::kMaxLocationLength;
+    char location[kMaxLen + 1] = { 0 };
+    size_t codeLen             = 0;
+
+    CHIP_ERROR err = ConfigurationMgr().GetCountryCode(location, sizeof(location), codeLen);
```

Comment:
Does GetCountryCode guarantee null-termination of the output?  Why is it also returning codeLen if it does?

I guess the existing code in the init method was making this assumption too....

Reply:
I did not check into that yet. The API is overall really poor :( It's mixed between null and non-null terminated strings in ConfigurationManager